test/pylib/s3_server_mock.py

```diff
@@ -0,0 +1,222 @@
+                                </CompleteMultipartUploadResult>"""
+                case Policy.RETRYABLE_FAILURE:
+                    # should succeed on retry
+                    self.policies.put(path, Policy.SUCCESS)
+                    return """<?xml version="1.0" encoding="UTF-8"?>
+
+                                <Error>
+                                    <Code>InternalError</Code>
+                                    <Message>We encountered an internal error. Please try again.</Message>
+                                    <RequestId>656c76696e6727732072657175657374</RequestId>
+                                    <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                </Error>"""
+                case Policy.NONRETRYABLE_FAILURE:
+                    return """<?xml version="1.0" encoding="UTF-8"?>
+
+                                <Error>
+                                    <Code>InvalidAction</Code>
+                                    <Message>Something went terribly wrong</Message>
+                                    <RequestId>656c76696e6727732072657175657374</RequestId>
+                                    <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                </Error>"""
+                case _:
+                    raise ValueError("Unknown policy")
+
+
+# Mock server to setup `ThreadingHTTPServer` instance with custom request handler (see above), managing requests state
+# in the `self.req_states`, adding custom logger, etc. This server will be started automatically from `test.py` once the
+# pytest kicks in. In addition, it is possible just to start this server by using another script - `start_mock.py` to
+# run it locally to provide mocking functionality for tests executed during development process
+class MockS3Server:
+    def __init__(self, host, port, logger=None):
+        self.req_states = LRUCache(10000)
+        handler = partial(InjectingHandler, self.req_states, logger)
+        self.server = ThreadingHTTPServer((host, port), handler)
+        self.server_thread = None
+        self.server.request_queue_size = 1000
+        self.server.timeout = 10000
+        self.server.socket.settimeout(10000)
+        self.server.socket.listen(1000)
+        self.is_running = False
+        os.environ['MOCK_S3_SERVER_PORT'] = f'{port}'
+        os.environ['MOCK_S3_SERVER_HOST'] = host
+
+    async def start(self):
+        if not self.is_running:
+            loop = asyncio.get_running_loop()
+            self.server_thread = loop.run_in_executor(None, self.server.serve_forever)
+            self.is_running = True
+            print(f'Starting S3 mock server on {self.server.server_address}')
+
+    async def stop(self):
+        if self.is_running:
+            self.server.shutdown()
+            await self.server_thread
+            self.is_running = False
+            print('Stopping S3 mock server')
```

Comment:
this message could be misleading, as the mock server is already stopped at this moment.

Reply:
Moved up